subcommands/keys/rotate_targets.go

```diff
@@ -0,0 +1,95 @@
+package keys
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+)
+
+func init() {
+	rotateTargets := &cobra.Command{
+		Use:   "rotate-targets <offline-creds.tgz>",
+		Short: "Rotate the offline target signing key for the Factory",
+		Run:   doRotateTargets,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			api = subcommands.Login(cmd)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	subcommands.RequireFactory(rotateTargets)
+	cmd.AddCommand(rotateTargets)
+}
+
+func doRotateTargets(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	credsFile := args[0]
+	assertWritable(credsFile)
+	creds := getOfflineCreds(credsFile)
+
+	root, err := api.TufRootGet(factory)
+	subcommands.DieNotNil(err)
+
+	// Target "rotation" works like this:
+	// 1. Find the "online target key" - this the key used by CI, so we don't
+	//    want to lose it.
+	// 2. Generate a new key.
+	// 3. Set these keys in root.json
+
+	onlineTargetId, err := findOnlineTargetId(factory, *root, creds)
+	subcommands.DieNotNil(err)
+
+	rootid, rootPk, err := findRoot(*root, creds)
+	fmt.Println("= Root keyid:", rootid)
+	subcommands.DieNotNil(err)
+	targetid, newCreds := replaceProdTargetKey(root, onlineTargetId, creds)
+	fmt.Println("= New target:", targetid)
+	subcommands.DieNotNil(signRoot(root, TufSigner{rootid, rootPk}))
+
+	tmpCreds := saveTempCreds(credsFile, newCreds)
+
+	bytes, err := json.MarshalIndent(root, "", "  ")
+	subcommands.DieNotNil(err)
+	fmt.Println("= Uploading new root")
+	body, err := api.TufRootPost(factory, bytes)
+	if err != nil {
+		fmt.Println("\nERROR:", err)
+		fmt.Println(body)
+		os.Exit(1)
+	}
+	if err := os.Rename(tmpCreds, credsFile); err != nil {
+		fmt.Println("\nERROR: Unable to update offline creds file.", err)
+		fmt.Println("Temp copy still available at:", tmpCreds)
+	}
+}
+
+func findOnlineTargetId(factory string, root client.AtsTufRoot, creds OfflineCreds) (string, error) {
+	onlinePub, err := api.GetFoundriesTargetsKey(factory)
+	subcommands.DieNotNil(err)
+	for _, keyid := range root.Signed.Roles["targets"].KeyIDs {
+		pub := root.Signed.Keys[keyid].KeyValue.Public
+		if pub == onlinePub.KeyValue.Public {
+			return keyid, nil
+		}
+	}
+	return "", errors.New("Unable to find online target key for factory")
+}
+
+func replaceProdTargetKey(root *client.AtsTufRoot, onlineTargetId string, creds OfflineCreds) (string, OfflineCreds) {
```

Comment:
Can you replace Prod with Offline everywhere ?
This patch set doesn't even create prod root role - it's CI root role, which has online and offline targets keys.

I'm scary we might overload a word "prod" rather quickly if we put it in every place which has something to do with production environment.

Reply:
good point: https://github.com/foundriesio/fioctl/pull/84/commits/26c454a725366fb61d10a83d015f7d95d6966a86

if that looks good, i'll squash fix ups and merge

Reply:
looks good